File: examples/spanner-client/ruby/cli_options.rb
# frozen_string_literal: true

CliOptions = Struct.new(:query_mode, :project, :instance, :database, :sql, keyword_init: true)

module CliOptionsParser
  VALID_MODES = %w[PLAN PROFILE].freeze
  DEFAULT_QUERY_FILE = File.expand_path('../query.sql', __dir__)

  module_function

  def usage
    <<~USAGE
      usage: analyze_and_render.rb [options]
        --query-mode PLAN|PROFILE   Spanner execute-sql mode (default: PLAN)
        --project PROJECT           GCP project id
        --instance INSTANCE         Spanner instance id
        --database DATABASE         Spanner database id
        --query SQL                 SQL text (overrides --query-file)
        --query-file PATH           SQL file (default: ../query.sql)

      Environment (when flags omitted):
        SPANNER_QUERY_MODE, SPANNER_PROJECT_ID, SPANNER_INSTANCE_ID,
        SPANNER_DATABASE_ID, SPANNER_QUERY, SPANNER_QUERY_FILE
    USAGE
  end

  def env_or_nil(name)
    value = ENV[name]
    return nil if value.nil? || value.strip.empty?

    value.strip
  end

  def load_sql(query, query_file)
    return query.strip unless query.nil? || query.strip.empty?

    path = query_file.nil? || query_file.strip.empty? ? DEFAULT_QUERY_FILE : query_file
    File.read(path, encoding: 'UTF-8').strip
  end

  def parse(argv = ARGV)
    query_mode = (env_or_nil('SPANNER_QUERY_MODE') || 'PLAN').upcase
    project = env_or_nil('SPANNER_PROJECT_ID')
    instance = env_or_nil('SPANNER_INSTANCE_ID')
    database = env_or_nil('SPANNER_DATABASE_ID')
    query = env_or_nil('SPANNER_QUERY')
    query_file = env_or_nil('SPANNER_QUERY_FILE')

    args = argv.dup
    i = 0
    while i < args.length
      arg = args[i]
      case arg
      when '-h', '--help'
        warn usage
        exit 0
      when '--query-mode'
        raise ArgumentError, 'missing value for --query-mode' unless i + 1 < args.length

        query_mode = args[i + 1].upcase
        i += 2
      when '--project'
        raise ArgumentError, 'missing value for --project' unless i + 1 < args.length

        project = args[i + 1]
        i += 2
      when '--instance'
        raise ArgumentError, 'missing value for --instance' unless i + 1 < args.length

        instance = args[i + 1]
        i += 2
      when '--database'
        raise ArgumentError, 'missing value for --database' unless i + 1 < args.length

        database = args[i + 1]
        i += 2
      when '--query'
        raise ArgumentError, 'missing value for --query' unless i + 1 < args.length

        query = args[i + 1]
        i += 2
      when '--query-file'
        raise ArgumentError, 'missing value for --query-file' unless i + 1 < args.length

        query_file = args[i + 1]
        i += 2
      else
        raise ArgumentError, "unknown argument: #{arg}"
      end
    end

    raise ArgumentError, "query mode must be PLAN or PROFILE, got: #{query_mode}" unless VALID_MODES.include?(query_mode)
    raise ArgumentError, 'missing required value: set --project or SPANNER_PROJECT_ID' if project.nil? || project.empty?
    raise ArgumentError, 'missing required value: set --instance or SPANNER_INSTANCE_ID' if instance.nil? || instance.empty?
    raise ArgumentError, 'missing required value: set --database or SPANNER_DATABASE_ID' if database.nil? || database.empty?

    CliOptions.new(
      query_mode: query_mode,
      project: project,
      instance: instance,
      database: database,
      sql: load_sql(query, query_file)
    )
  end
end

File: examples/spanner-client/python/cli_options.py
"""Parse CLI flags and environment for Spanner client examples."""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from pathlib import Path

_VALID_MODES = frozenset({"PLAN", "PROFILE"})
_DEFAULT_QUERY_FILE = Path(__file__).resolve().parents[1] / "query.sql"


@dataclass(frozen=True)
class CliOptions:
    query_mode: str
    project: str
    instance: str
    database: str
    sql: str


def _usage() -> str:
    return (
        "usage: analyze_and_render.py [options]\n"
        "  --query-mode PLAN|PROFILE   Spanner execute-sql mode (default: PLAN)\n"
        "  --project PROJECT           GCP project id\n"
        "  --instance INSTANCE         Spanner instance id\n"
        "  --database DATABASE         Spanner database id\n"
        "  --query SQL                 SQL text (overrides --query-file)\n"
        "  --query-file PATH           SQL file (default: ../query.sql)\n"
        "\n"
        "Environment (when flags omitted):\n"
        "  SPANNER_QUERY_MODE, SPANNER_PROJECT_ID, SPANNER_INSTANCE_ID,\n"
        "  SPANNER_DATABASE_ID, SPANNER_QUERY, SPANNER_QUERY_FILE"
    )


def _env_or_none(name: str) -> str | None:
    value = os.environ.get(name)
    if value is None or not value.strip():
        return None
    return value.strip()


def _load_sql(query: str | None, query_file: str | None) -> str:
    if query is not None:
        return query.strip()
    path = Path(query_file) if query_file else _DEFAULT_QUERY_FILE
    return path.read_text(encoding="utf-8").strip()


def parse_cli_options(argv: list[str] | None = None) -> CliOptions:
    args = list(argv if argv is not None else sys.argv[1:])
    query_mode = (_env_or_none("SPANNER_QUERY_MODE") or "PLAN").upper()
    project = _env_or_none("SPANNER_PROJECT_ID")
    instance = _env_or_none("SPANNER_INSTANCE_ID")
    database = _env_or_none("SPANNER_DATABASE_ID")
    query = _env_or_none("SPANNER_QUERY")
    query_file = _env_or_none("SPANNER_QUERY_FILE")

    i = 0
    while i < len(args):
        arg = args[i]
        if arg in ("-h", "--help"):
            print(_usage(), file=sys.stderr)
            raise SystemExit(0)
        if arg == "--query-mode" and i + 1 < len(args):
            query_mode = args[i + 1].upper()
            i += 2
            continue
        if arg == "--project" and i + 1 < len(args):
            project = args[i + 1]
            i += 2
            continue
        if arg == "--instance" and i + 1 < len(args):
            instance = args[i + 1]
            i += 2
            continue
        if arg == "--database" and i + 1 < len(args):
            database = args[i + 1]
            i += 2
            continue
        if arg == "--query" and i + 1 < len(args):
            query = args[i + 1]
            i += 2
            continue
        if arg == "--query-file" and i + 1 < len(args):
            query_file = args[i + 1]
            i += 2
            continue
        raise SystemExit(f"unknown argument: {arg}")

    if query_mode not in _VALID_MODES:
        raise SystemExit(f"query mode must be PLAN or PROFILE, got: {query_mode}")
    if not project:
        raise SystemExit("missing required value: set --project or SPANNER_PROJECT_ID")
    if not instance:
        raise SystemExit("missing required value: set --instance or SPANNER_INSTANCE_ID")
    if not database:
        raise SystemExit("missing required value: set --database or SPANNER_DATABASE_ID")

    return CliOptions(
        query_mode=query_mode,
        project=project,
        instance=instance,
        database=database,
        sql=_load_sql(query, query_file),
    )
